Update SampleBase.cpp

selimsandal opened the pull request:
Since window name is get before actually selecting the API we display the wrong window name, also we can't set VULKAN from commandline even though help shows we should be able to: `graphics API: D3D11, D3D12 or VULKAN (string [=D3D12])`. This fixes the problem determining API before window creation.

selimsandal commented:
Also I see that d3d12 does not do vsync by default but vulkan does, should I look into that or is it intended?

selimsandal commented:
also with d3d12 sceneviewer has severe flickering but with vulkan I can't see it, maybe because the FPS limit I can't tell right now

dzhdanNV commented:
`apiName == selectedApi == cmdLine.get<std::string>("api")`: looks like your changes actually do nothing, if I understand correctly.

> Also I see that d3d12 does not do vsync by default but vulkan does, should I look into that or is it intended?

It's NRI framework unrelated. All samples do:
```c++
swapChainDesc.flags = (m_Vsync ? nri::SwapChainBits::VSYNC : nri::SwapChainBits::NONE) | nri::SwapChainBits::ALLOW_TEARING;
```
But if your VK device supports the only one `VK_PRERSENT_MODE_FIFO_KHR` - that's it, it's Vsync = ON. Please, check in the debugger what happens inside swap chain creation around this code block:
```c++
VkPresentModeKHR presentMode = VK_PRESENT_MODE_FIFO_KHR; // the only one 100% supported (see spec)
{
 ...
}
```

> also with d3d12 sceneviewer has severe flickering but with vulkan I can't see it, maybe because the FPS limit I can't tell right now

Which GPU? What is flickering: UI, scene, entire frame?

dzhdanNV commented:
> apiName == selectedApi == cmdLine.get<std::string>("api"): looks like your changes actually do nothing, if I understand correctly.

Sorry, I have understood. Your changes fix the case when "api" is not set. I will do a better fix tomorrow. Thanks for the idea!

selimsandal commented:
> Which GPU? What is flickering: UI, scene, entire frame?

RTX 5090, latest hotfix drivers 576.66, will send a vid not at my pc right now. But happens on another GPU too I don't think it's specific to my device.

dzhdanNV commented:
OK. Thanks. Trying on RTX 4080...

This is a proper fix (preview ^_^):
![image](https://github.com/user-attachments/assets/2e5dbf42-5087-48ed-9cd3-8e5d9f60d2a3)

dzhdanNV commented:
> OK. Thanks. Trying on RTX 4080...

No flickering. Latest driver for me is 576.66. I will try to repro in another way, but need to know how it looks like. No rush. I can wait...

selimsandal commented:
> This is a proper fix (preview ^_^):

didn't see that api to string function, much better yeah :)

https://github.com/user-attachments/assets/d9a8889a-4e04-44e9-ad61-c8d2d5d64307

here is a video I got using remote desktop, not very high res but you get the idea. Left half of the screen is more highres and flickers, right half is blurry / more pixelated. Same thing happens on a 3070 for me.

dzhdanNV commented:
Ouch! It's expected. `SceneViewer` is not only a simple test for drawing commands. It also:
- does pixel jittering via PSL (programmable sample locations)
- right part of the screen uses VRS (variable rate shading)

Basically, `SceneViewer` is a unit test for 3 features in one sample app :)

selimsandal commented:
Oh got it, didn't look at the code closely since I can't see the effect on my mac. Shouldn't it happen on mac too? Is the screen too high res for me to see that?

dzhdanNV commented:
> Shouldn't it happen on mac too?

Features are enabled only if supported. So, I suspect no support for PSL and VRS on Mac.

dzhdanNV commented:
To conclude this, please choose one of:
- close this PR
- feel free to implement the code from my screenshot and update the PR

Thanks for your interest!

selimsandal commented:
> > Shouldn't it happen on mac too?
> 
> Features are enabled only if supported. So, I suspect no support for PSL and VRS on Mac.

Got it, makes sense. One more thing, I wrote some of the examples from scratch using SDL3 and DirectXShaderCompiler in a standalone project that does not use NRIFramework. NRISamples using that structure prevented me from starting quickly for my engine. I think it would be nice if we can host a standalone sample that only uses NRI as a direct dependency. Can help people. If you want to do that I can send the sample. Does that make sense?

I'll update the PR quickly

selimsandal commented:
updated PR, also fixed cmake config error on macos, if a subproject enables OBJC but not top project cmake has a bug that prevents objc compiler from being found.

https://gitlab.kitware.com/cmake/cmake/-/issues/24104

dzhdanNV commented:
>  I think it would be nice if we can host a standalone sample that only uses NRI as a direct dependency. Can help people. If you want to do that I can send the sample. Does that make sense?

NRI framework is even not a framework. It's some minimal glue code hiding window and input management and GLTF loading. NRI samples could be abstracted even more, but they are not because we wanted (nowadays I want) to keep them explicit as much as possible. But at the same time not crossing "the line of explicitness". At this moment I would prefer not to add new entities. Moreover, SDL3 is just another window/input API (+more). I use GLFW. If your ask can be rephrased to "we need more console samples using only NRI". It's doable. `Buffers` and `DeviceInfo` use raw NRI (but they are console apps). Of course, if you think that this is important feel free to create a new issue in NRI samples, I think. And again, I very appreciate your interest and involvement.

dzhdanNV commented:
Good one. Merged! Thank you!

selimsandal commented:
> NRI framework is even not a framework. It's some minimal glue code hiding window and input management and GLTF loading. NRI samples could be abstracted even more, but they are not because we wanted (nowadays I want) to keep them explicit as much as possible. But at the same time not crossing "the line of explicitness". At this moment I would prefer not to add new entities. Moreover, SDL3 is just another window/input API (+more). I use GLFW. If your ask can be rephrased to "we need more console samples using only NRI". It's doable. `Buffers` and `DeviceInfo` use raw NRI (but they are console apps). Of course, if you think that this is important feel free to create a new issue in NRI samples, I think. And again, I very appreciate your interest and involvement.

No problem, thanks for the explanation. About more samples, I think the current basic samples are great as starting points and I managed to bootstrap my engine with NRI quickly using them. It would be great to have more advanced samples like NRD Sample that integrates other NVIDIA SDKs, (like MegaGeometry or RTXNS) if you have the time to write them.

Thanks!

dzhdanNV commented:
I got your point.
